### PR TITLE
Add LAMB optimizer support

### PR DESCRIPTION
**Pull request by xwli-chelsea**
# Description

Add support to use LAMB optimizer. See paper [Large Batch Optimization for Deep Learning: Training BERT in 76 minutes](https://arxiv.org/abs/1904.00962).

This implementation is similar to the official implementation from tf addon https://www.tensorflow.org/addons/api_docs/python/tfa/optimizers/LAMB

Note that we does not apply adam bias correction for the moments estimates to keep it similar to the original AdamW optimizer in BERT pretraining.

Fixes # (issue)

## Type of change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)

## List all changes 
Please list all changes in the commit.
* Added the LAMBOptimizer class and option to switch from bert_adam and bert_lamb. 
* Added test
* Updated param list

# Testing

pytest
tested with run_detext.sh on toy dataset

**Test Configuration**:
* Firmware version:
* Hardware:
* Toolchain:
* SDK:

# Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: src/detext/train/optimization.py [hunk 1]
```diff
@@ -66,10 +66,14 @@ def create_optimizer(hparams, loss):
...
+        name_2_optimizer = {
+            'bert_adam': AdamWeightDecayOptimizer,
+            'bert_lamb': LAMBOptimizer
+        }
         # It is recommended that you use this optimizer for fine tuning, since this
-        # is how the model was trained (note that the Adam m/v variables are NOT
+        # is how the model was trained (note that the Adam/Lamb m/v variables are NOT
         # loaded from init_checkpoint.)
-        optimizer = AdamWeightDecayOptimizer(
+        optimizer = name_2_optimizer[hparams.optimizer](
```
[the remark this answers is not in the document]
Comment (xwli-chelsea) [reply src/detext/train/optimization.py:76]: sounds good.

File: src/detext/train/optimization.py [hunk 2]
```diff
@@ -235,3 +239,105 @@ def _get_variable_name(self, param_name):
...
         return param_name
+
+
+class LAMBOptimizer(tf.train.Optimizer):
+    """
+    Optimizer that implements the Layer-wise Adaptive Moments (LAMB).
+    See paper [Large Batch Optimization for Deep Learning: Training BERT
+        in 76 minutes](https://arxiv.org/abs/1904.00962).
+    Official implementation from tensorflow addon tensorflow_addons/optimizers/lamb.py
+    Note that we does not apply adam bias correction for the moments estimates to keep it similar to the original AdamW
```
[the remark this answers is not in the document]
Comment (xwli-chelsea) [reply src/detext/train/optimization.py:250]: in the pre-training experiments, we observed noticeable convergence slow down by adding the bias correction. Therefore it's kept this way.

File: src/detext/train/optimization.py [hunk 3]
```diff
@@ -235,3 +239,105 @@ def _get_variable_name(self, param_name):
...
+                initializer=tf.zeros_initializer())
+            v = tf.get_variable(
+                name=param_name + "/lamb_v",
+                shape=param.shape.as_list(),
+                dtype=tf.float32,
+                trainable=False,
+                initializer=tf.zeros_initializer())
+
+            # Standard Adam update.
+            next_m = (tf.multiply(self.beta_1, m) + tf.multiply(1.0 - self.beta_1, grad))
```
[the remark this answers is not in the document]
Comment (xwli-chelsea) [reply src/detext/train/optimization.py:295]: there's no difference.

File: setup.py [hunk 4]
```diff
@@ -15,7 +15,7 @@
                  "Intended Audience :: Developers",
                  "License :: OSI Approved"],
     license='BSD-2-CLAUSE',
-    version='1.1.5',
+    version='1.1.7',
```
Comment (zhoutong-fu) [inline setup.py:18]: Not an issue but I saw this version needs to be updated constantly. Maybe we can consider automating it later.

File: src/detext/run_detext.py [hunk 5]
```diff
@@ -71,7 +71,7 @@ def add_arguments(parser):
                         help="Whether to use normalized lm. This option only works for unit_type=lstm_lm")
 
     # Optimizer
-    parser.add_argument("--optimizer", type=str, choices=["sgd", "adam", "bert_adam"], default="sgd",
+    parser.add_argument("--optimizer", type=str, choices=["sgd", "adam", "bert_adam", "bert_lamb"], default="sgd",
```
Comment (zhoutong-fu) [inline src/detext/run_detext.py:74]: As we have more choices for the optimizer, it'll be great if we can have a section on the optimizers in the training instruction

File: src/detext/train/optimization.py [hunk 6]
```diff
@@ -235,3 +239,105 @@ def _get_variable_name(self, param_name):
...
             param_name = m.group(1)
         return param_name
+
+
+class LAMBOptimizer(tf.train.Optimizer):
+    """
+    Optimizer that implements the Layer-wise Adaptive Moments (LAMB).
+    See paper [Large Batch Optimization for Deep Learning: Training BERT
+        in 76 minutes](https://arxiv.org/abs/1904.00962).
+    Official implementation from tensorflow addon tensorflow_addons/optimizers/lamb.py
```
[the remark this answers is not in the document]
Comment (xwli-chelsea) [reply src/detext/train/optimization.py:249]: The tf addon optimizers inherit from tf.keras.optimizers.Optimizer. helper functions like `_resource_apply_dense`, `_resource_apply_sparse`, `_create_slots`  and `_create_slots are overriding` the parent class methods. The maths are still the same.

File: setup.py (hunk 4 above)
Comment (xwli-chelsea) [reply setup.py:18]: Yes it'll be nice to have it automated. There are solutions but haven't got a chance to pull in.

File: src/detext/run_detext.py (hunk 5 above)
Comment (xwli-chelsea) [reply src/detext/run_detext.py:74]: I see. Do you mean in the TRAINING.md? I can add some reference for the different optimizers in the parameter table.

File: src/detext/train/optimization.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (StarWang) [reply src/detext/train/optimization.py:250]: It's better we open the authentic LAMB optimizer to users.
Can we add an option to control whether we want to apply bias correction? In this way we can let users choose whether they want a fast or authentic LAMB

[the remark this answers is not in the document]
Comment (xwli-chelsea) [reply src/detext/train/optimization.py:250]: The bias correction is not something specific to the LAMB optimizer. It's related to Adam update. The only part that's different than the AdamW implementation in the optimization code is the layer adaptation part.

[the remark this answers is not in the document]
Comment (StarWang) [reply src/detext/train/optimization.py:250]: Thanks for the explanation! Then I'm good whether bias correction is added or not.
Can you detail the reasoning in the comment a bit as you've explained in this thread?

[the remark this answers is not in the document]
Comment (xwli-chelsea) [reply src/detext/train/optimization.py:250]: 👌  sure. will add something.

[the remark this answers is not in the document]
Comment (xwli-chelsea) [reply src/detext/train/optimization.py:258]: updated in this commit: https://github.com/linkedin/detext/pull/21/commits/4ee62de6abc4e0be4216866e650e351b7a52cd83

File: src/detext/train/optimization.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (StarWang) [reply src/detext/train/optimization.py:256]: nit: `optimizer_func` is actually a class and you may want to use capital casing

File: src/detext/train/optimization.py [hunk 7]
```diff
@@ -75,10 +75,12 @@ def create_optimizer(hparams, loss):
             'bert_adam': AdamWeightDecayOptimizer,
             'bert_lamb': LAMBOptimizer
         }
+        optimizer_func = name_2_optimizer[hparams.optimizer]
```
Comment (StarWang) [inline src/detext/train/optimization.py:78]: nit: optimizer_func is actually a class and you may want to use capital casing

Comment (StarWang) [reply src/detext/train/optimization.py]: LGTM now. Thanks!

**Pull request merged**